1_Download_seqs.R

```diff
@@ -0,0 +1,146 @@
+                                              rettype="gb",
+                                              retmode = "text")
+        gb_info_list[[j]] <- gb_info_temp
+        recs <- rentrez::entrez_fetch(db="nuccore",
+                                      web_history=upload,
+                                      rettype="fasta") # if restez package is loaded need to define rentrez package here
+        cat(recs, file=paste0(path_to_output_dir, search_term, ".fasta"), append=TRUE)
+        cat("\n",seq_start[j]+chunk-1, "specimens info has been fetched\n")
+      }
+      if(j == batch_n){
+        final_add <- max_seq - seq_start[j]
+        upload <- entrez_post(db = "nuccore",
+                              id = Caption[seq_start[j]:(seq_start[j]+final_add)])
+        gb_info_temp <- rentrez::entrez_fetch(db="nuccore",
+                                              web_history = upload,
+                                              rettype="gb",
+                                              retmode = "text")
+        gb_info_list[[j]] <- gb_info_temp
+        recs <- rentrez::entrez_fetch(db="nuccore",
+                                      web_history=upload,
+                                      rettype="fasta") # if restez package is loaded need to define rentrez package here
+        cat(recs, file=paste0(path_to_output_dir, search_term, ".fasta"), append=TRUE)
+        cat("\n", max_seq, "specimens info has been fetched\n")
+      }
+    }
+    
+    gb_info <- paste0(unlist(gb_info_list), collapse = "")
+    gb_info_list2 <- strsplit(gb_info, split = "(?<=LOCUS)", perl = TRUE) # split along LOCUS, but keep LOCUS in result too
+    
+    # I need to refer to the 1st element of the list where I have all the sequences. I need to skip the first among the sequences which is just one word: LOCUS
+    feature <- lapply(2:length(gb_info_list2[[1]]), function(x) gb_extract(record = gb_info_list2[[1]][x], what = "features")[[1]])
+    acc <- lapply(2:length(gb_info_list2[[1]]), function(x) gb_extract(record = gb_info_list2[[1]][x], what = "accession")[[1]])
+    extra_info <- unique(unlist(lapply(feature, names)))
+    all_info <- matrix(ncol = length(extra_info))
+    for(i in 1:length(feature)){
+      n <- unlist(feature[[i]])
+      names(n) <- names(feature[[i]])
+      n2 <- n[match(extra_info, names(n))] # with this I expand the vector to the sice of extra_info, NA will be for no matches
+      # names(n2) <- extra_info
+      all_info <- rbind(all_info, n2)
+    }
+    all_info <- all_info[-1,]
+    colnames(all_info) <- extra_info
+    all_info2 <- as.data.frame(all_info)
+    all_info2$accession_no <- unlist(acc)
+    if(nrow(meta_keep) == nrow(all_info2)){
+      df2 <- cbind(meta_keep, all_info2)
+    }else{
+      cat("You couldn't download all the sequences, try to increase the chunk size")
+    }
+
+    write.table(x = df2, file = paste0(path_to_output_dir, search_term, ".metadata.tsv"),
```

Comment:
Currently this throws an error if not all sequences were downloaded (as `df2` never gets created in that scenario). You could either put `write.table` within the ifelse above, or figure out how to make an intermediate df2 with not all downloaded sequences to write to file in that scenario (although maybe that's not so useful anyway?)

Reply:
That is what I was afraid of.. the entrez_post function could have a bug, or its performance depends on the network. I don't know... how many sequences did you aim to download? Have you tried to increase the value of the chunk argument?

Reply:
Yeah it's quite variable, sometimes I get them all fine and sometimes some are missing. To be honest it seems to be independent of chunk size. For now I think the most conservative thing is to put `write.table` inside the if statement, and try and come up with a solution for matching the metadata to the number of sequences *actually downloaded* later - from a quick look something like extracting the fasta headers from the `rentrez::entrez_fetch` string output during the download for loop, and then matching to the captions to filter `meta_keep`

Reply:
Ok, then, this is not a viable direction. I have already checked the rentrez GitHub and others have similar problems.  The best would be to avoid the post function for which we should be able to edit the web history that I tried previously, but I had the same problem like this: https://github.com/ropensci/rentrez/issues/144

Unfortunately, if the metadata (all_info2) does not match with the meta_keep, then it has already missed the sequences from the fasta files too. So in my opinion what we can do is:

1.  Download all sequences and filter the data after it: we can avoid the  entrez_post function, which I think is the main problem, and Emily and Brigid had a similar problem with it. As soon as I bypassed this function, I got consistent results (see the previous version of the script)
2. Emily wrote a for loop to download sequences multiple times: I think it is not viable for big data
3. If I understand well, your suggestion is to check the IDs on the fly, so during the for loops. Actually, this gave me the idea that I could just check the missing IDs using the metadata tables and run a while loop with only the few sequence IDs that are skipped for some reason until all the desired IDs are in the extended meta table. It is similar to what Emily did, but not repeating everything, just those that we are missing!
4. If the IDs could be appended to a web history object as mentioned above, the problem would be solved

I will try to go for option 3 in the following days because even with a while loop, it would be more efficient than downloading everything and then filtering (in some cases, there could be thousands of wgs data), and probably the first attempt will miss only just a few sequences. What do you think?